### PR TITLE
Fix #150. Add run configuration for Function App Publishing

### LINKED ISSUE #150: Functions: Deploy
Similar to web app deployment?

### PR DESCRIPTION
**Pull request by sdubov**
- Add new run configuration for Function App Publishing
- Split more components
- Move base logic between Web and Function Apps into a base classes
- Add validation for a Function App form
- Add function app context provider for publish Function App
- Fix App Service Plan validation message for invalid length
- Update plugin description

### CODE REVIEW

File: PluginsAndFeatures/azure-toolkit-for-intellij/rider/resources/META-INF/plugin.xml [hunk 1]
```diff
@@ -27,8 +27,13 @@
               <li>Create SQL Database from Azure Explorer (<a href="https://github.com/JetBrains/azure-tools-for-intellij/pull/135">#135</a>)</li>
             </ul>
           </li>
+          <li>Azure Function Apps
+            <ul>
+              <li>Publish Function Apps with run configuration or from a project explorer context menu (<a href="https://github.com/JetBrains/azure-tools-for-intellij/pull/150">#150</a>)</li>
+            </ul>
```
Comment (maartenba) [inline PluginsAndFeatures/azure-toolkit-for-intellij/rider/resources/META-INF/plugin.xml:35]: May also need something like "Integrates with Azure Functions Core Tools integration: project and item templates, and a run configuration to run/debug .NET Core function apps"

File: PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/com/microsoft/intellij/runner/functionapp/config/FunctionAppConfigurationType.kt [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+import com.intellij.execution.configurations.ConfigurationType
+import com.intellij.openapi.util.IconLoader
+import org.jetbrains.annotations.Nls
+import javax.swing.Icon
+
+class FunctionAppConfigurationType : ConfigurationType {
+
+    companion object {
+        private const val RUN_CONFIG_TYPE_ID = "AzureFunctionAppPublish"
+        private const val RUN_CONFIG_TYPE_NAME = "Azure Publish Function App"
```
Comment (maartenba) [inline PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/com/microsoft/intellij/runner/functionapp/config/FunctionAppConfigurationType.kt]: Maybe "Azure - Publish Function App" or "Publish Azure Function app"? (and same for the web app publish configuration)

Looks kind of weird now:

image

**Review by maartenba — APPROVED** [review]
Two minor things that could be good to fix.

File: PluginsAndFeatures/azure-toolkit-for-intellij/rider/resources/META-INF/plugin.xml (hunk 1 above)
Comment (sdubov) [reply PluginsAndFeatures/azure-toolkit-for-intellij/rider/resources/META-INF/plugin.xml:35]: Add info about run/debug and templates. Please feel free to re-phrase.

File: PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/com/microsoft/intellij/runner/functionapp/config/FunctionAppConfigurationType.kt (hunk 2 above)
Comment (sdubov) [reply PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/com/microsoft/intellij/runner/functionapp/config/FunctionAppConfigurationType.kt]: Fixed.

**Pull request merged**